system/src/system_control_internal.cpp

```diff
@@ -81,30 +81,49 @@ SystemControl g_systemControl;
 SystemControl::SystemControl() :
 #ifdef USB_VENDOR_REQUEST_ENABLE
         usbChannel_(this),
-#endif
-#if HAL_PLATFORM_BLE
-        bleChannel_(this),
 #endif
         appReqHandler_(nullptr) {
 }
 
 int SystemControl::init() {
 #if HAL_PLATFORM_BLE
-    const int ret = bleChannel_.init();
-    if (ret != 0) {
-        return ret;
+    // BleControlRequestChannel::init() should be called only once
+    // as characteristic UUIDs are currently called only once. If BLE prov mode
+    // is enabled, we need to make sure to call characteristic UUIDs in prov
+    // mode system calls for the first time instead of here.
+    // See system_ble_prov_mode()
+    if (!HAL_Feature_Get(FEATURE_DISABLE_LISTENING_MODE)) {
+        const int ret = BleControlRequestChannel::instance(this)->init();
+        if (ret != 0) {
+            return ret;
+        }
     }
 #endif
     return 0;
 }
 
 void SystemControl::run() {
 #if HAL_PLATFORM_BLE
-    bleChannel_.run();
+    BleControlRequestChannel::instance(this)->run();
 #endif
 }
 
 void SystemControl::processRequest(ctrl_request* req, ControlRequestChannel* /* channel */) {
+    // All requests are allowed if vector is empty
+    // If vector is non-empty, only those requests in the vector are allowed. Others are filtered out
+    if (!vecCtrlReq_.isEmpty()) {
```

Comment:
There should be an easy way to disable all requests. A special request type (SOMETHING_SOMETHING_ANY) can be used.